src/backend/distributed/commands/alter_table.c

```diff
@@ -0,0 +1,1297 @@
+	con->cascadeToColocated = params->cascadeToColocated;
+
+	Relation relation = try_relation_open(con->relationId, ExclusiveLock);
+	if (relation == NULL)
+	{
+		ereport(ERROR, (errmsg("cannot complete operation "
+							   "because no such table exists")));
+	}
+	relation_close(relation, NoLock);
+	con->distributionKey =
+		BuildDistributionKeyFromColumnName(relation, con->distributionColumn);
+
+	con->originalAccessMethod = NULL;
+#if PG_VERSION_NUM >= PG_VERSION_12
+	if (!PartitionedTable(con->relationId))
+	{
+		HeapTuple amTuple = SearchSysCache1(AMOID, ObjectIdGetDatum(
+												relation->rd_rel->relam));
+		if (!HeapTupleIsValid(amTuple))
+		{
+			ereport(ERROR, (errmsg("cache lookup failed for access method %d",
+								   relation->rd_rel->relam)));
+		}
+		Form_pg_am amForm = (Form_pg_am) GETSTRUCT(amTuple);
+		con->originalAccessMethod = NameStr(amForm->amname);
+		ReleaseSysCache(amTuple);
+	}
+#endif
+
+
+	con->colocatedTableList = NIL;
+	if (IsCitusTableType(con->relationId, DISTRIBUTED_TABLE))
+	{
+		con->originalDistributionKey = DistPartitionKey(con->relationId);
+
+		CitusTableCacheEntry *cacheEntry = GetCitusTableCacheEntry(con->relationId);
+		con->originalShardCount = cacheEntry->shardIntervalArrayLength;
+
+		List *colocatedTableList = ColocatedTableList(con->relationId);
+
+		/*
+		 * we will not add partition tables to the colocatedTableList
+		 * since they will be handled separately.
+		 */
+		Oid colocatedTableId = InvalidOid;
+		foreach_oid(colocatedTableId, colocatedTableList)
+		{
+			if (PartitionTable(colocatedTableId))
+			{
+				continue;
+			}
+			con->colocatedTableList = lappend_oid(con->colocatedTableList,
+												  colocatedTableId);
+		}
+	}
+
+	/* find relation and schema names */
+	con->relationName = get_rel_name(con->relationId);
+	con->schemaId = get_rel_namespace(con->relationId);
+	con->schemaName = get_namespace_name(con->schemaId);
+
+	/* calculate a temp name for the new table */
+	con->tempName = pstrdup(con->relationName);
+	con->hashOfName = hash_any((unsigned char *) con->tempName, strlen(con->tempName));
+	AppendShardIdToName(&con->tempName, con->hashOfName);
+
+	if (con->conversionType == UNDISTRIBUTE_TABLE)
+	{
+		con->function = &UndistributeTable;
+	}
+	if (con->conversionType == ALTER_DISTRIBUTED_TABLE)
+	{
+		con->function = &AlterDistributedTable;
+	}
+	if (con->conversionType == ALTER_TABLE_SET_ACCESS_METHOD)
+	{
+		con->function = &AlterTableSetAccessMethod;
+	}
+
+	return con;
+}
+
+
+void
+CreateDistributedTableLike(TableConversionState *con)
+{
+	Var *newDistributionKey =
+		con->distributionColumn ? con->distributionKey : con->originalDistributionKey;
+
+	char *newColocateWith = con->colocateWith;
+	if (con->colocateWith == NULL)
+	{
+		if (con->originalDistributionKey->vartype == newDistributionKey->vartype &&
+			con->shardCountIsNull)
+		{
+			newColocateWith =
+				quote_qualified_identifier(con->schemaName, con->relationName);
```

Comment:
I'm a bit confused by this. Are we co-locating the table with itself? This could use a comment.